pkg/util/netlink.go

```diff
@@ -0,0 +1,135 @@
+package util
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/vishvananda/netlink"
+)
+
+func ModeFromString(s string) (netlink.MacvlanMode, error) {
+	switch s {
+	case "", "bridge":
+		return netlink.MACVLAN_MODE_BRIDGE, nil
+	case "private":
+		return netlink.MACVLAN_MODE_PRIVATE, nil
+	case "vepa":
+		return netlink.MACVLAN_MODE_VEPA, nil
+	default:
+		return 0, fmt.Errorf("unknown macvtap mode: %q", s)
+	}
+}
+
+func CreateMacvtap(name string, master string, mode string) (int, error) {
+	// attempt to delete any previously existing link
+	if l, _ := netlink.LinkByName(name); l != nil {
+		_ = netlink.LinkDel(l)
+	}
+
+	m, err := netlink.LinkByName(master)
+	if err != nil {
+		return 0, fmt.Errorf("failed to lookup master %q: %v", master, err)
+	}
+
+	nlmode, err := ModeFromString(mode)
+	if err != nil {
+		return 0, err
+	}
+
+	mv := &netlink.Macvtap{
+		Macvlan: netlink.Macvlan{
+			LinkAttrs: netlink.LinkAttrs{
+				Name:        name,
+				ParentIndex: m.Attrs().Index,
+				// we had crashes if we did not set txqlen to some value
+				TxQLen: m.Attrs().TxQLen,
+			},
+			Mode: nlmode,
+		},
+	}
+
+	if err := netlink.LinkAdd(mv); err != nil {
+		return 0, fmt.Errorf("failed to create macvtap: %v", err)
+	}
+
+	if err := netlink.LinkSetUp(mv); err != nil {
+		return 0, fmt.Errorf("failed to set %q UP: %v", name, err)
+	}
+
+	return mv.Attrs().Index, nil
+}
+
+func RecreateMacvtap(name string, master string, mode string) (int, error) {
+	err := LinkDelete(name)
+	if err != nil {
+		return 0, err
+	}
+	return CreateMacvtap(name, master, mode)
+}
+
+func LinkExists(link string) error {
```

Comment:
This should return bool. If it fails, it does not necessarily mean the link does not exist. Could you check the returned value? As you do in the following function.